### PR TITLE
New SumStream function reads from io.Reader

### PR DESCRIPTION
**Pull request by gammazero**
This provides a way to read from a stream instead of having to load a potentially large file into memory first.

Addresses issue #15

### CODE REVIEW

File: sum.go [hunk 1]
```diff
@@ -13,17 +15,26 @@ var ErrSumNotSupported = mhreg.ErrSumNotSupported
 var ErrLenTooLarge = errors.New("requested length was too large for digest")
 
 // Sum obtains the cryptographic sum of a given buffer. The length parameter
-// indicates the length of the resulting digest and passing a negative value
-// use default length values for the selected hash function.
+// indicates the length of the resulting digest. Passing a negative value uses
+// default length values for the selected hash function.
 func Sum(data []byte, code uint64, length int) (Multihash, error) {
+	return SumStream(bytes.NewReader(data), code, length)
```
Comment (warpfork) [inline sum.go:21]: I'm afraid we should take the hit of some duplicated code here, in exchange for keeping this a single `[]byte` handoff without reader indirections.  We'd rather not add one or more allocations to all code which currently uses this function with smallish byte slices.

**Review by warpfork — CHANGES_REQUESTED** [review]
Unfortunately I find myself asking for slightly less DRY code.  But otherwise this LGTM!

File: sum.go (hunk 1 above)
Comment (gammazero) [reply sum.go]: Refactored to write `[]byte` directly without wrapping in `bytes.Reader`, with minimal code dup.

**Review by warpfork — APPROVED** [review]

**Comment by warpfork** [comment]
Thanks for adding this!  It really connects the dots well.

**Pull request merged**